src/WebJobs.Script/Description/DotNet/Compilation/CSharp/CachedCSharpCompilation.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public sealed class CachedCSharpCompilation : ICSharpCompilation
+    {
+        private readonly ICSharpCompilation _innerCompilation;
+        private readonly FunctionMetadata _functionMetadata;
+        private readonly ILogger _logger;
+        private static readonly ConcurrentDictionary<string, Task<Assembly>> _cachedCompilations = new ConcurrentDictionary<string, Task<Assembly>>();
+
+        public CachedCSharpCompilation(ICSharpCompilation compilation, FunctionMetadata functionMetadata, ILoggerFactory loggerFactory)
+        {
+            _innerCompilation = compilation;
+            _functionMetadata = functionMetadata;
+            _logger = loggerFactory?.CreateLogger(LogCategories.Startup) ?? NullLogger.Instance;
```

Comment:
We compile outside of startup as well, so I'm not sure this is the right log category.